### ISSUE TITLE
Using deep_transform_keys(&:to_sym on unsafe strong parameters hash gives keys as Strings

### ISSUE THREAD

**Issue by rgisiger** [rails/rails#47406]
First of all, I don't know if this is really an issue, but I can't find any notes about this change, and it was working differently in version 6.1.x of Rails.

### Steps to reproduce

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails", branch: "7-0-stable"
end

require "action_controller/railtie"
require "active_support"
require "active_support/core_ext/object/blank"
require "minitest/autorun"

class BugTest < Minitest::Test
  def test_stuff
      params = ActionController::Parameters.new(a: "123", b: "456")
      assert params.to_unsafe_h.deep_symbolize_keys.keys.all?(Symbol)
      assert params.to_unsafe_h.deep_transform_keys(&:to_sym).keys.all?(Symbol)
  end
end
```

### Expected behavior
A Hash with keys as symbols should be returned

### Actual behavior
An ActiveSupport::HashWithIndifferentAccess with keys as strings is returned

### System configuration
**Rails version**: 7.0.x

**Ruby version**: 2.7.7

**Comment by adrianna-chang-shopify** [comment]
I don't think this is a bug, but rather the expected behaviour for `ActiveSupport::HashWithIndifferentAccess`.
From the [docs](https://api.rubyonrails.org/classes/ActiveSupport/HashWithIndifferentAccess.html):
> Internally symbols are mapped to strings when used as keys in the entire writing interface (calling []=, merge, etc). This mapping belongs to the public interface. For example, given:

> hash = ActiveSupport::HashWithIndifferentAccess.new(a: 1)
> You are guaranteed that the key is returned as a string:

> hash.keys # => ["a"]

I believe the change in behaviour came from https://github.com/rails/rails/pull/40703, because `HashWithIndifferentAccess#deep_transform_keys` was previously (incorrectly) returning a `Hash`, but now returns the appropriate `HashWithIndifferentAccess`, whose keys are internally mapped to strings.

**Comment by rgisiger** [comment]
Ok I see the point now, it's clearly not a bug.
Thank you very much for the research, I can close this then.
